File: PAT (Advanced Level) Practice/1003 Emergency.py
# This is the best version I can figure out with python
# Accepted on Cases 1-5 TLE on Case 6
# This algorithm is fine
# io takes too long time
# cpp version passes all cases, see in the same folder

from collections import deque

class Citys:
    def __init__(self, size):
        self.size = size
        self.paths = [[float('inf') for i in range(size)] for i in range(size)]
        self.teams = [0 for i in range(size)]
        self.src = -1
        self.dst = -1
        self.minimium = float('inf')
        self.mans = 0
        self.num = 0

    def set_teams(self):
        self.teams = [int(i) for i in input().split()]


    def set_path(self, times):
        for _ in range(times):
            a, b, v = [int(i) for i in input().split()]
            #self.paths[a][b] = v
            self.paths[b][a] = v


    def dfs(self, used, paths, mans):
        if (used and used[-1] == self.dst):
            if self.minimium > paths:
                self.num = 1
                self.minimium = paths
                self.mans = mans
            elif self.minimium == paths:
                self.num += 1
                if self.mans < mans:
                    self.mans = mans
            return
        else:
            if used:
                for each in set(range(self.size)) - set(used):
                    temp = self.paths[used[-1]][each] + paths
                    if temp == float('inf') or self.minimium < temp:
                        continue
                    if temp <= self.minimium:
                        used.append(each)
                        self.dfs(used, temp, mans + self.teams[each])
                        used.pop()
            else:
                used.append(self.src)
                self.dfs(used, paths , mans + self.teams[self.src])
                



n, m, a, b = [int(i) for i in input().split()]
city = Citys(n)
city.set_teams()


for _ in range(m):
    aa, bb, v = [int(i) for i in input().split()]
    city.paths[aa][bb] = v
    city.paths[bb][aa] = v
city.src = a
city.dst = b

city.dfs(deque(),0,0)

print('%d %d' % (city.num, city.mans))

File: PAT (Advanced Level) Practice/1003 Emergency.cpp
#include <cstdio>
#include <vector>
#define INF 65535
using namespace std;


// Cpp version for 1003 Emergency.py
// Accpeted on all 6 cases

class Citys{
    public:
    vector<vector<int>> map;    
    vector<int> teams;
    int size;
    int minimum = INF;
    int mans = 0;
    int src = -1;
    int dst = -1;
    int num = 0;
    vector<int> used;

    void init(){
        int size;
        int lines;
        scanf("%d %d %d %d", &size, &lines, &this->src, &this->dst);
        this->set_size(size);
        this->set_teams();
        this->set_edge(lines);
    }

    bool find(int v){
        for(int i = 0; i < this->used.size(); ++i){
            if(this->used[i] == v){
                return true;
            }
        }
        return false;
    }

    void set_size(int size){
        this->size = size;
        this->teams.resize(size);
        this->map.resize(size);
        for(int i=0; i<size; i++){
            this->map[i].resize(size);
        }
        for(int i=0; i<size; i++){
            for(int j=0; j<size; j++){
                map[i][j] = INF;
            }
        }
    }

    void set_edge(int m){
        int a, b, v;
        for(int i = 0; i < m; ++i){
            scanf("%d %d %d", &a, &b, &v);
            this->map[a][b] = v;
            this->map[b][a] = v;
        }

    }

    void set_teams(){
        int team;
        for(int i = 0; i < this->size; ++i){
            scanf("%d", &team);
            this->teams[i] = team;
        }
    }

    void dfs(int paths, int mans){
        if(!this->used.empty()){
            if(this->used[this->used.size()-1] == this->dst){
                if(this->minimum > paths){
                    this->num = 1;
                    this->minimum = paths;
                    this->mans = mans;
                }else if (this->minimum == paths){
                    this->num++;
                    if(this->mans < mans){
                        this->mans = mans;
                    }
                }
                return ;
            }
        }
        {
            if(!this->used.empty()){
                int temp;
                for(int each = 0; each < this->size; each++){
                    if(find(each)) continue;
                    temp = this->map[this->used[this->used.size()-1]][each];
                    if(temp >= INF) continue;
                    temp += paths;
                    if( this->minimum < temp) continue;
                    if(temp <= this->minimum){
                        used.push_back(each);
                        this->dfs(temp, mans + this->teams[each]);
                        used.pop_back();
                    }
                }
            }
            else{
                used.push_back(this->src);
                this->dfs(paths , mans + this->teams[this->src]);
            }
        }
    }

    
    
};

int main(){
    Citys city;
    city.init();
    city.dfs(0, 0);
    printf("%d %d", city.num, city.mans);
}